pkg/controller/upload-controller.go

```diff
@@ -0,0 +1,440 @@
+		// delete everything
+
+		// delete service
+		err = c.deleteService(pvc.Namespace, resourceName)
+		if err != nil {
+			return errors.Wrapf(err, "Error deleting upload service for pvc: %s", key)
+		}
+
+		// delete pod
+		err = c.deletePod(pvc.Namespace, resourceName)
+		if err != nil {
+			return errors.Wrapf(err, "Error deleting upload pod for pvc: %s", key)
+		}
+
+		return nil
+	}
+
+	podPhaseFromPVC := func(pvc *v1.PersistentVolumeClaim) string {
+		phase, _ := pvc.ObjectMeta.Annotations[AnnUploadPodPhase]
+		return phase
+	}
+
+	podSucceededFromPVC := func(pvc *v1.PersistentVolumeClaim) bool {
+		return (podPhaseFromPVC(pvc) == string(v1.PodSucceeded))
+	}
+
+	if !podSucceededFromPVC(pvc) {
+		pod, err := c.getOrCreateUploadPod(pvc, resourceName)
+		if err != nil {
+			return errors.Wrapf(err, "Error creating upload pod for pvc: %s", key)
+		}
+
+		podPhase := string(pod.Status.Phase)
+		if podPhase != podPhaseFromPVC(pvc) {
+			var labels map[string]string
+			annotations := map[string]string{AnnUploadPodPhase: podPhase}
+			pvc, err = updatePVC(c.clientset, pvc, annotations, labels)
+			if err != nil {
+				return errors.Wrapf(err, "Error updating pvc %s, pod phase %s", key, podPhase)
+			}
+		}
+	}
+
+	if podSucceededFromPVC(pvc) {
+		// delete service
+		if err = c.deleteService(pvc.Namespace, resourceName); err != nil {
+			return errors.Wrapf(err, "Error deleting upload service for pvc %s", key)
+		}
+	} else {
+		// make sure the service exists
+		if _, err = c.getOrCreateUploadService(pvc, resourceName); err != nil {
+			return errors.Wrapf(err, "Error getting/creating service resource for PVC %s", key)
+		}
+	}
+
+	return nil
+}
+
+func (c *UploadController) getOrCreateUploadPod(pvc *v1.PersistentVolumeClaim, name string) (*v1.Pod, error) {
+	pod, err := c.podLister.Pods(pvc.Namespace).Get(name)
+
+	if apierrs.IsNotFound(err) {
+		pod, err = CreateUploadPod(c.clientset, c.serverCAKeyPair, c.clientCAKeyPair.Cert, c.uploadServiceImage, c.verbose, c.pullPolicy, name, pvc)
+	}
+
+	if pod != nil && !metav1.IsControlledBy(pod, pvc) {
+		return nil, errors.Errorf("%s pod not controlled by pvc %s", name, pvc.Name)
+	}
+
+	return pod, err
+}
+
+func (c *UploadController) getOrCreateUploadService(pvc *v1.PersistentVolumeClaim, name string) (*v1.Service, error) {
+	service, err := c.serviceLister.Services(pvc.Namespace).Get(name)
+
+	if apierrs.IsNotFound(err) {
+		service, err = CreateUploadService(c.clientset, name, pvc)
+	}
+
+	if service != nil && !metav1.IsControlledBy(service, pvc) {
+		return nil, errors.Errorf("%s service not controlled by pvc %s", name, pvc.Name)
+	}
+
+	return service, err
+}
+
+func (c *UploadController) deletePod(namespace, name string) error {
+	_, err := c.podLister.Pods(namespace).Get(name)
+	if apierrs.IsNotFound(err) {
+		return nil
+	}
+	if err == nil {
+		err = c.clientset.CoreV1().Pods(namespace).Delete(name, &metav1.DeleteOptions{})
```

Comment:
Since you aren't using expectations, check the error here. If the error is "IsNotFound" the ignore the error.